src/main/scala/org/constellation/util/MajorityStateChooser.scala

```diff
@@ -3,55 +3,100 @@ package org.constellation.util
+      case (recSnap, count) => recSnap
+    }
+    if (simpleMaj.isDefined) simpleMaj
+    else if (totalPeers == numNodesWithSnapAtThisHeight) { //todo need actual logic to get majority from a distribution. This could be improved
+      val votedMajority
+        : Option[RecentSnapshot] = snapVotes.sortBy { case (recSnap, count) => (count, recSnap.hash) }.map {
+        case (count, hash) => count
+      }.headOption
+      votedMajority
+    } else None
+  }
+
+  def getAllSnapsUntilMaj(snapshotNodes: SnapshotNodes, nodeSnapshots: List[NodeSnapshots]) = {
+    val (nodeWithMajority, allSnapshots): (Id, Seq[RecentSnapshot]) = nodeSnapshots.filter {
+      case (id, snaps) => snapshotNodes._2.contains(id)
+    }.maxBy(snaps => snaps._2.length)
+    allSnapshots.filter(_.height <= snapshotNodes._1.height)//todo filter such that, download peers contain current node's
+    // snap/height or rather there is an intersection without missing heights
+  }
 
-  def chooseMajorityState(
-    nodeSnapshots: List[NodeSnapshots],
-    ownHeight: Long,
-    allPeers: Seq[Id]
-  ): OptionT[F, (Seq[RecentSnapshot], Set[Id])] =
-    for {
-      majorState <- chooseMajoritySnapshot(
-        nodeSnapshots.filter(checkIfNodeContainsSnapshotsInConsistentState),
-        ownHeight,
-        allPeers
-      )
-      snapsThoughMaj <- OptionT.fromOption[F](getAllSnapsUntilMaj(majorState._2, nodeSnapshots))
-//      node <- OptionT.fromOption[F](findNode(nodeSnapshots, nodeId))
-      nodeIds <- OptionT.fromOption[F](chooseMajNodeIds(majorState._2, nodeSnapshots))
-
-      _ <- OptionT.liftF(logger.debug(s"Re-download from nodes : ${nodeIds}"))
-    } yield (snapsThoughMaj.sortBy(-_.height), nodeIds.toSet)
-
-  private def chooseMajoritySnapshot(nodeSnapshots: Seq[NodeSnapshots], height: Long,
-                                     allPeers: Seq[Id]) =
-    for {
-      // highestSnapshot <- OptionT.fromOption[F](getHighest(nodeSnapshots))
-      // useHighest <- OptionT.liftF(shouldUseHighest(highestSnapshot, ownHeight))
-      majorSnapshot <- OptionT.fromOption[F](chooseMajorWinner(allPeers, nodeSnapshots))//def chooseMajorWinner(allPeers: Seq[Id])(nodeSnapshots: Seq[NodeSnapshots])
-
-      _ <- OptionT.liftF(
-        logger.debug(
-//          s"The highest snapshot : $highestSnapshot"
-          s"The major snapshot: $majorSnapshot"
-        )
-      )
-    } yield majorSnapshot
+  def chooseMajorWinner(allPeers: Seq[Id], nodeSnapshots: Seq[NodeSnapshots]) = {
+    val snapshotToProposers: Seq[(RecentSnapshot, Seq[(Id, RecentSnapshot)])] = nodeSnapshots.flatMap {
+      case (id, snapshots) => snapshots.map((id, _))
+    }.groupBy {
+      case (_, snapshot) => snapshot
+    }.toList
+    val snapsGroupedByHeight: Map[Long, Seq[(RecentSnapshot, Seq[Id])]] = snapshotToProposers.map {
+      case (recentSnap, idRecentSnapTups) => (recentSnap, idRecentSnapTups.map(_._1))
+    }.groupBy { case (recentSnap, idRecentSnaps) => recentSnap.height }
+    val res = Try {
```

Comment:
Why you're using `Try` and then `toOption`. Can't you write it as an option from the beginning?

Reply:
We should change this, its because .maxBy is used below which errors on empty collections. We should copy the sortBy.headOption above in selectMajSnap